pkg/kubelet/metrics/collectors/emptydir_metrics.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collectors
+
+import (
+	"context"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/component-base/metrics"
+	"k8s.io/klog/v2"
+
+	kubeletmetrics "k8s.io/kubernetes/pkg/kubelet/metrics"
+	serverstats "k8s.io/kubernetes/pkg/kubelet/server/stats"
+)
+
+var (
+	emptyDirUsedBytesDesc = metrics.NewDesc(
+		metrics.BuildFQName(
+			"",
+			kubeletmetrics.KubeletSubsystem,
+			kubeletmetrics.EmptyDirUsedBytesKey,
+		),
+		"Bytes used by the emptyDir volume.",
+		[]string{
+			"volume_name",
+			"namespace",
+			"pod",
+		},
```

Comment:
The cardinility should be approximately O(ephemeralStorage). Since we have two metrics, it should be two times that amount.

There was some discussions in the past in https://github.com/kubernetes/kubernetes/issues/69507 that suggested that in practice this should be reasonnable, but still we should be careful since this is unbounded and the churn could be big considering the ephemeral nature of these volumes.

@jsafrane do you perhaps know if there are real world scenarios where hundred of thousands of ephemeral storage are created and these new metrics could blow up monitoring platforms?

Reply:
(Note that https://github.com/kubernetes/kubernetes/blob/bbd83d86444d7b325a51f4daa0d65163b795b70e/pkg/kubelet/metrics/collectors/volume_stats.go for PVC related metrics could be considered as a precedent, particularly in the context of local persistent storage.)

Reply:
If really only EmptyDir with `StorageMediumDefault` are reported, then I would expect that nr. of such volumes is smaller than nr. of Pods. And we reported bunch of per-pod metrics last time I checked.